src/lib/device/journal_vdev.cpp

```diff
@@ -34,75 +34,211 @@
+    m_end_offset += new_chunk->size();
+
+    if (!m_journal_chunks.empty()) {
+        // If there are already chunks in the m_journal_chunks list, append this new chunk to the end of the list. Write
+        // the next_chunk of the last chunk in the list to point to this new chunk. If already there are no chunks make
+        // the new chunk as the head.
+        auto last_chunk = m_journal_chunks.back();
+        auto* last_chunk_private = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(last_chunk->user_private()));
+
+        // Set the next chunk with the newly created chunk id.
+        last_chunk_private->next_chunk = new_chunk->chunk_id();
+
+        // Append the new chunk
+        m_journal_chunks.push_back(new_chunk);
+        auto chunk_size = m_vdev.info().chunk_size;
+        auto offset_in_chunk = (tail_offset() % chunk_size);
+        if (offset_in_chunk != 0) {
+            // Update the overhead to total write size
+            m_write_sz_in_total.fetch_add(last_chunk->size() - offset_in_chunk, std::memory_order_relaxed);
+            last_chunk_private->end_of_chunk_size = offset_in_chunk;
         }
+        m_vdev.update_chunk_private(last_chunk, last_chunk_private);
+        HS_LOG(TRACE, device, "desc : {} added chunk new {} previous : {}", m_logdev_id, new_chunk->chunk_id(),
+               last_chunk->chunk_id());
     } else {
-        // across chunk boundary and no space left;
-        HS_LOG(ERROR, device, "No space left! m_write_sz_in_total: {}, m_reserved_sz: {}", m_write_sz_in_total.load(),
-               m_reserved_sz);
-        return INVALID_OFFSET;
-        // m_reserved_sz stays sthe same;
+        // If the list is empty, update the new chunk as the head. Only head chunk contains the logdev_id.
+        auto* new_chunk_private = r_cast< JournalChunkPrivate* >(const_cast< uint8_t* >(new_chunk->user_private()));
+        new_chunk_private->is_head = true;
+        new_chunk_private->logdev_id = m_logdev_id;
+        // Append the new chunk
+        m_journal_chunks.push_back(new_chunk);
+        m_vdev.update_chunk_private(new_chunk, new_chunk_private);
+        HS_LOG(TRACE, device, "desc : {} added head chunk {}", m_logdev_id, new_chunk->chunk_id());
+    }
+}
+
+off_t JournalVirtualDev::Descriptor::alloc_next_append_blk(size_t sz) {
+    // We currently assume size requested is less than chunk_size.
+    auto chunk_size = m_vdev.info().chunk_size;
+    RELEASE_ASSERT_LT(sz, chunk_size, "Size requested greater than chunk size");
+
+    if ((tail_offset() + static_cast< off_t >(sz)) >= m_end_offset) {
+        // not enough space left, add a new chunk.
+        HS_LOG(DEBUG, device,
+               "No space left, desc: {} m_write_sz_in_total: {}, m_reserved_sz: {} tail: 0x{}. Creating chunk",
+               m_logdev_id, m_write_sz_in_total.load(), m_reserved_sz, to_hex(tail_offset()));
+
+#ifdef _PRERELEASE
+        iomgr_flip::test_and_abort("abort_before_update_eof_cur_chunk");
```

Comment:
Are these flips enabled and running fine somewhere?

Reply:
We need to add it in our long running tests